File: src/client/token-management.ts
#! /usr/bin/env ts-node

/*
 * Description: Example of using the Keeta Network Client to Create and Manage a Token.
 */

import * as KeetaNet from '@keetanetwork/keetanet-client';
import { debugPrintableObject as DPO, getFaucetTokens } from '../helper.js';
import * as util from 'util';

const Account = KeetaNet.lib.Account;
const network = 'test';

async function main() {
	// Generate a random seed and user accounts using different KeyAlgorithm
	const seed = Account.generateRandomSeed({ asString: true });
	const userAccount = Account.fromSeed(seed, 0);
	// Create UserClient for Network and User Account
	await using userClient = KeetaNet.UserClient.fromNetwork(network, userAccount);

	// Get Tokens from the Faucet for Fees
	const faucetRequest = await getFaucetTokens(userAccount, network);

	if (!faucetRequest) {
		throw(new Error('Failed to get Faucet Tokens'));
	}

	// Create a Token on the Network
	const { account: token } = await userClient.generateIdentifier(Account.AccountKeyAlgorithm.TOKEN);
	if (!token.isToken()) {
		// basic type narrowing for AccountKeyAlgorithm.TOKEN
		throw(new Error('Tokens Should be TOKEN Key Algorithm'));
	}

	// Metadata like number of decimals places can be stored in the token account info
	const basicMetadata = Buffer.from(JSON.stringify({ decimalPlaces: 10 }), 'utf-8').toString('base64');

	// Initialize a Builder to batch multiple operations
	const builder = userClient.initBuilder();
	builder.setInfo({
		name: 'TKNA',
		description: `Example Token`,
		metadata: basicMetadata,
		// Default Permissions apply to every transaction using this token.  ACCESS grants everyone access to the token
		defaultPermission: new KeetaNet.lib.Permissions(['ACCESS'])
		// Because the builder was initialized with the user account we need to specify which account want to set the info on
	}, { account: token });

	// Setup Initial Token Supply - we use 50_000 * number of decimals/precision to get the raw amount
	builder.modifyTokenSupply(50_000n * (10n ** 10n), { account: token });

	/**
	 * Compute blocks to ensure token supply is added before sending.
	 * The builder does not always guarantee correct order so we compute the supply blocks to ensure they come first
	 */
	await builder.computeBlocks();

	// Send 200 TKNA Supply to user account
	builder.send(userAccount, 200n * (10n ** 10n), token, undefined, { account: token });

	// Compute Final Blocks
	await builder.computeBlocks();

	console.log(`User Seed: ${seed}`);
	console.log(`User Account: ${userAccount.publicKeyString.get()}`);
	console.log(`$TKNA Token: ${token.publicKeyString.get()}`);
	console.log(`Token Setup Blocks: ${[builder.blocks.map(block => block.hash.toString())]}`);

	// Publish builder blocks to the network
	await builder.publish();

	// Get final balances to compare after swap
	const userBalance = await userClient.allBalances();
	const tokenBalance = await userClient.client.getAllBalances(token);
	const tokenInfo = await userClient.client.getAccountInfo(token);

	console.log(`Initial User Balance: ${util.inspect(DPO(userBalance), { depth: 4, colors: true })}`);
	console.log(`Token Balance: ${util.inspect(DPO(tokenBalance), { depth: 4, colors: true })}`);
	console.log(`Token Info: ${util.inspect(DPO(tokenInfo), { depth: 4, colors: true })}`);
}

main().then(function() {
	process.exit(0);
}, function(err: unknown) {
	console.error(err);
	process.exit(1);
});

File: csharp/src/Ledger/TokenManagement.cs
using System.Numerics;
using System.Text;
using System.Text.Json;
using KeetaNet.Anchor;
using KeetaNet.Anchor.Crypto;
using KeetaNet.Examples.Common;

namespace KeetaNet.Examples.Ledger;

public sealed class TokenManagementExample : IKeetaExample
{
	private const KeetaNetwork Network = KeetaNetwork.Test;
	private const int DecimalPlaces = 10;

	public string Id => "client/token-management";

	public string Description =>
		"Example of using the Keeta Network Client to Create and Manage a Token";

	public async Task<int> Run(string[] args, CancellationToken cancellationToken = default)
	{
		using var runtime = WasmRuntime.Load();

		// Generate a random seed and user account.
		string seed = runtime.Accounts.GenerateRandomSeed();
		using Account userAccount = runtime.Accounts.FromSeed(seed, 0, "ecdsa_secp256k1");
		using UserClient userClient = runtime.CreateUserClient(Network, userAccount);

		// Get tokens from the faucet for fees.
		if (!await Helper.GetFaucetTokens(userClient, Network, cancellationToken))
		{
			throw new InvalidOperationException("Failed to get Faucet Tokens");
		}

		// Create a token on the network.
		using Account token = await userClient.GenerateIdentifier(
			IdentifierKind.Token, cancellationToken: cancellationToken);

		// Metadata like the number of decimal places is stored in the token
		// account info.
		string basicMetadata = Convert.ToBase64String(
			Encoding.UTF8.GetBytes(JsonSerializer.Serialize(new { decimalPlaces = DecimalPlaces })));

		BigInteger scale = BigInteger.Pow(10, DecimalPlaces);

		// The token setup block runs on the token's own chain, signed by the
		// owner. ACCESS as the default permission grants everyone access to
		// the token.
		using Permissions access = runtime.Blocks.PermissionsFromFlags(new[] { BaseFlag.Access });
		using BlockOperation setInfo = runtime.Blocks.SetInfo(
			"TKNA", "Example Token", basicMetadata, access);
		using BlockOperation supply = runtime.Blocks.TokenAdminSupply(50_000 * scale, AdjustMethod.Add);

		using Block setupBlock = BuildTokenBlock(runtime, userAccount, token, previous: null, setInfo, supply);

		// Send 200 TKNA of the fresh supply from the token to the user, chained
		// atop the setup block.
		using BlockOperation send = runtime.Blocks.Send(userAccount, 200 * scale, token);
		using Block sendBlock = BuildTokenBlock(runtime, userAccount, token, setupBlock.Hash, send);

		Console.WriteLine($"User Seed: {seed}");
		Console.WriteLine($"User Account: {userAccount.PublicKeyString}");
		Console.WriteLine($"$TKNA Token: {token.PublicKeyString}");
		Console.WriteLine($"Token Setup Blocks: {setupBlock.Hash}, {sendBlock.Hash}");

		// Publish both blocks to the network. The user pays any demanded fee.
		bool published = await userClient.Client.Transmit(
			new[] { setupBlock, sendBlock },
			TransmitOptions.WithFeeSigner(userAccount),
			cancellationToken);
		if (!published)
		{
			throw new InvalidOperationException("Failed to publish the token setup blocks");
		}

		// Read the final state back from the network.
		IReadOnlyList<TokenBalance> userBalances = await userClient.GetAllBalances(cancellationToken);
		IReadOnlyList<TokenBalance> tokenBalances = await userClient.Client.GetAllBalances(token, cancellationToken);
		AccountState tokenInfo = await userClient.Client.GetAccountInfo(token, cancellationToken);

		Console.WriteLine($"User Balances: {FormatBalances(userBalances)}");
		Console.WriteLine($"Token Balances: {FormatBalances(tokenBalances)}");
		Console.WriteLine($"Token Info: {tokenInfo.Info?.Name} ({tokenInfo.Info?.Description}), supply {tokenInfo.Info?.Supply}");

		return 0;
	}

	/// <summary>Build one signed block on the token's chain with the given operations.</summary>
	private static Block BuildTokenBlock(
		WasmRuntime runtime,
		Account owner,
		Account token,
		BlockHash? previous,
		params BlockOperation[] operations)
	{
		using BlockBuilder builder = runtime.Blocks.NewBuilder();
		builder
			.WithVersion(2)
			.WithNetwork(Network.Id())
			.WithAccount(token)
			.WithSigner(owner)
			.WithDate(DateTimeOffset.UtcNow);

		if (previous is { } hash)
		{
			builder.WithPrevious(hash);
		}
		else
		{
			builder.AsOpening();
		}

		foreach (BlockOperation operation in operations)
		{
			builder.AddOperation(operation);
		}

		return builder.Build();
	}

	private static string FormatBalances(IReadOnlyList<TokenBalance> balances) =>
		string.Join(", ", balances.Select(entry => $"{entry.Token.PublicKeyString}={entry.Balance}"));
}
